src/autolabel/transforms/webpage_transform.py

```diff
@@ -0,0 +1,115 @@
+import asyncio
+import logging
+import pandas as pd
+import ssl
+
+logger = logging.getLogger(__name__)
+
+MAX_RETRIES = 3
+MAX_KEEPALIVE_CONNECTIONS = 20
+MAX_CONNECTIONS = 100
+BACKOFF = 2
+HEADERS = {}
+HTML_PARSER = "html.parser"
+
+
+class WebpageTransform(BaseTransform):
+    def __init__(
+        self,
+        url_column: str,
+        output_columns: List[str],
+        timeout: int = 5,
+        headers: Dict[str, str] = HEADERS,
+    ) -> None:
+        super().__init__(output_columns)
+        self.url_column = url_column
+        self.headers = headers
+        self.max_retries = MAX_RETRIES
+        try:
+            from bs4 import BeautifulSoup
+            import httpx
+
+            if not headers.get("User-Agent"):
+                from fake_useragent import UserAgent
+
+                headers["User-Agent"] = UserAgent().random
+
+            self.httpx = httpx
+            self.timeout = httpx.Timeout(timeout)
+            limits = httpx.Limits(
+                max_keepalive_connections=MAX_KEEPALIVE_CONNECTIONS,
+                max_connections=MAX_CONNECTIONS,
+                keepalive_expiry=timeout,
+            )
+            self.client = httpx.AsyncClient(
+                timeout=self.timeout, limits=limits, follow_redirects=True
+            )
+            self.client_with_no_verify = httpx.AsyncClient(
+                timeout=self.timeout, limits=limits, follow_redirects=True, verify=False
+            )
+            self.beautiful_soup = BeautifulSoup
+        except ImportError:
+            raise ImportError(
+                "BeautifulSoup, httpx and fake_useragent are required to use the webpage transform. Please install them with the following command: pip install bs4 httpx fake_useragent"
+            )
+
+    def name(self) -> str:
+        return TransformType.WEBPAGE_TRANSFORM
+
+    def _load_metadata(self, url, soup) -> Dict[str, Any]:
+        metadata = {"url": url}
+        if soup.find("title"):
+            metadata["title"] = soup.find("title").get_text()
+        for meta in soup.find_all("meta"):
+            if meta.get("name") and meta.get("content"):
+                metadata[meta.get("name")] = meta.get("content")
+            elif meta.get("property") and meta.get("content"):
+                metadata[meta.get("property")] = meta.get("content")
+        return metadata
+
+    async def _load_url(
+        self, url: str, verify=True, headers=HEADERS, retry_count=0
+    ) -> Tuple[str, Dict[str, Any]]:
+        if retry_count >= self.max_retries:
+            logger.warning(f"Max retries reached for URL: {url}")
+            return None, None
+
+        try:
+            client = self.client
+            if not verify:
+                client = self.client_with_no_verify
+            response = await client.get(url, headers=headers)
+
+            # TODO: Add support for other parsers
+            soup = self.beautiful_soup(response.text, HTML_PARSER)
```

Comment:
It make sense to add parsers as part of the config and then the init of the transform will handle instantiating the parser.

However, in the future we might want to have a parsing class which can parse the webpage as you want and we define the usual parsers, i.e text parser, url parser, table parser etc. In that case even this might be passed as a class. However, that is difficult to do in the config ways of doing things and registering parsers seems to be out of scope for now

Edit: (Adding parsing as a separate transform is also possible however we need html in order to transform, extracting urls may not be possible with the text format)

Reply:
I agree we can keep them as two separate set of transforms. The webpage transform just loads a given url and returns the loaded html, text or soup. The parser in this context refers to the beautiful soup parsers such as html.parser, lxml and html5lib.
The other transform support extracting content/urls from a raw html, bytes or soup.
That way each of these transforms have separation of concerns and be extended independent of each other.